### ISSUE TITLE
Why does being invariant to image-level operations help style transfer learning?

### ISSUE THREAD

**Issue by MLSlayer** [junyanz/pytorch-CycleGAN-and-pix2pix#872]
@phillipi mentioned that
'instance norm will be (nearly) invariant to image-level operations like changing the exposure or contrast of a photo, whereas batchnorm will not. Batchnorm is only invariant to batch-level operations.'

but why does that help in style transfer learning for cycleGan?

**Comment by phillipi** [comment]
Good point, for style transfer you might not want to be invariant to things like overall brightness -- a bright input should generate a bright stylized output.

But most of the action of these normalization layers is happening deep in the networks, on feature maps. So it's hard to know exactly what invariances they are actually achieving -- the brightness example is just to provide an intuition of what it would be like to apply instance norm to the raw pxels themselves.

There are a variety of ideas for why normalization helps in general in deep nets. I'm not sure a clear story has emerged, but it may have to do with [covariate shift](https://arxiv.org/abs/1502.03167), or it may be more about [smoothing the optimization landscape](https://arxiv.org/abs/1805.11604), or perhaps it's more about regularization (limiting the capacty of networks by making them invariant to certain nuisances).

**Comment by junyanz** [comment]
BTW, we don't use the normalization layer in the first layer, so that it can still preserve the brightness or other low-level statistics of the input image to some degree.

**Comment by MLSlayer** [comment]
@phillipi Thanks for your feedback! I suppose the details are not so so important to practitioners, as we just have to know that in most cases, normalization will help not hurt. By the way, for Batchnorm at least, the latest story/idea is that covariate shift is only a small chance part of it and it has more to do with making the optimization process easier by adding the gamma and beta parameters. They help be able to tune the activation fed into batchnorm based on just those two params in a linear like faction, w(x) + b, where w is gamma and b is beta, whereas the original activation might have a lot more params that work in more complex/intricate ways.  I mean, there is just a bunch of more theory that can be said about normalization, all of which you mentioned is plausible. I was just hoping someone knew the definite answer and can provide a concrete example for instancenorm in the case of why it specifically helps style gans, but intuitions are great!

Also for InstanceNorm, look at this quote from the paper https://arxiv.org/abs/1607.08022

It says about instancenorm that 

'This prevents instance-specific mean and covariance shift simplifying the learning process.'

just to reinforce your intuition.

Lastly, for batchnorm vs instancenorm, I'm sure you already know this but, the gamma and beta  I mentioned isn't the case when you are using the default params for InstanceNorm2d since affine=False means no learnable params. So most likely, it does help with all three others instead, covariate shift, making the cost function easier to optimize by reducing the elongation of the cost function and also adding some regularization.


@junyanz 
Now, Jun Yan, I believe that isn't the case for cycleGan. Take a look at your code here for the resnet9 option,

```
class ResnetGenerator(nn.Module):
    def __init__(self, input_nc, output_nc, ngf=64, norm_layer=nn.BatchNorm2d, use_dropout=False, n_blocks=6, padding_type='reflect'):
        assert(n_blocks >= 0)
        super(ResnetGenerator, self).__init__()
        if type(norm_layer) == functools.partial:
            use_bias = norm_layer.func == nn.InstanceNorm2d
        else:
            use_bias = norm_layer == nn.InstanceNorm2d

        model = [nn.ReflectionPad2d(3),
                 nn.Conv2d(input_nc, ngf, kernel_size=7, padding=0, bias=use_bias),
                 norm_layer(ngf),
                 nn.ReLU(True)]

```

you initialize your model, with instancenorm included in the first layer. Perhaps you were referring to a different model? 

However, I do appreciate extra feedback such as those. Those little details about knowing when to stay away from normalization is super nice.

**Comment by junyanz** [comment]
Sorry for the confusion. I mean that we didn't use the normalization in the first layer of the discriminator. If the low-level statistics of output need to be related to those of the input, the discriminator has a chance to enforce it.

**Comment by MLSlayer** [comment]
Ah, thanks for the clarification! Great idea!
